File: Scala/graph_algorithms/src/main/scala/minimum_spanning_trees/OptimalClustering.scala
package minimum_spanning_trees

object OptimalClustering {
  import scala.annotation.tailrec

  final case class Point(x: Double, y: Double)

  final case class PointPairDistance(indexOfP: Int, indexOfQ: Int, distance: Double)

  class UnionFind(nrPoints: Int) {
    private val parentIndices: Array[Int] = (0 until nrPoints).toArray
    private val ranks: Array[Int] = Array.fill(nrPoints)(0)

    private def changeParentsToRoot(indicesOnPath: List[Int], root: Int): Unit =
      indicesOnPath.foreach(ix => parentIndices(ix) = root)

    def find(childIndex: Int): Int = {
      @tailrec
      def loop(id: Int, parentId: Int, indicesTowardsRoot: List[Int]): (Int, List[Int]) = {
        if (id == parentId) (id, indicesTowardsRoot)
        else loop(parentId, parentIndices(parentId), id :: indicesTowardsRoot)
      }
      val (root, indicesOnPath): (Int, List[Int]) = loop(childIndex, parentIndices(childIndex), Nil)
      changeParentsToRoot(indicesOnPath, root)
      root
    }

    def union(parentIndexP: Int, parentIndexQ: Int): Unit = {
      if (parentIndexP != parentIndexQ) {
        if (ranks(parentIndexP) > ranks(parentIndexQ)) parentIndices(parentIndexQ) = parentIndexP
        else {
          parentIndices(parentIndexP) = parentIndexQ
          if (ranks(parentIndexP) == ranks(parentIndexQ)) ranks(parentIndexQ) += 1
        }
      }
    }
  }

  private def convertToDoubleList(line: String): List[Double] = line.split(" ").map(_.toDouble).toList

  private def calcDistance(p: Point, q: Point): Double = math.hypot(p.x - q.x, p.y - q.y)

  private def calcPairwisedistances(points: List[Point]): List[PointPairDistance] = {
    val distances: List[PointPairDistance] = (for {
      (p, ix) <- points.view.zipWithIndex
      (q, jy) <- points.drop(ix + 1).view.zipWithIndex
    } yield PointPairDistance(ix, ix + jy + 1, calcDistance(p, q))).toList
    distances.sortBy { case PointPairDistance(_, _, dist) => dist }
  }

  def calcOptimalClustering(nrPoints: Int, points: List[Point], k: Int): Double = {
    val distances: List[PointPairDistance] = calcPairwisedistances(points)
    val clusters = new UnionFind(nrPoints)

    @tailrec
    def loop(nrClusters: Int, xs: List[PointPairDistance]): Double = {
      val PointPairDistance(ixP, ixQ, dist) = xs.head
      val clusterOfP: Int = clusters.find(ixP)
      val clusterOfQ: Int = clusters.find(ixQ)
      if (clusterOfP == clusterOfQ) loop(nrClusters, xs.tail)
      else {
        clusters.union(clusterOfP, clusterOfQ)
        if (nrClusters == k) dist
        else loop(nrClusters - 1, xs.tail)
      }
    }

    loop(nrPoints, distances)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrPoints: Int = reader.next().toInt
    val points: List[Point] = (for { _ <- 0 until nrPoints } yield {
      val List(x, y): List[Double] = convertToDoubleList(reader.next())
      Point(x, y)
    }).toList
    val k: Int = reader.next().toInt
    val d: Double = calcOptimalClustering(nrPoints, points, k)
    println(d)
  }
}

File: OCaml/graph_algorithms/minimum_spanning_trees/optimal_clustering.ml
open Batteries

type point = { x : float; y: float }
type point_pair_distance = { index_of_p : int; index_of_q : int; distance : float }

module type UnionFindSig =
    sig
        type t
        val initialize : int -> t
        val find : t -> int -> int
        val union : t -> int -> int -> unit
    end

module UnionFind : UnionFindSig =
    struct
        type t = { parents : int array; ranks : int array }

        let change_parents_to_root parents indices_on_path root =
            List.iter (function ix -> parents.(ix) <- root) indices_on_path

        let initialize nr_points =
            { parents = Array.init nr_points (function ix -> ix); ranks = Array.make nr_points 0 }

        let find { parents } child_ix =
            let rec loop id parent_id indices_towards_root =
                if id = parent_id then id, indices_towards_root
                else loop parent_id parents.(parent_id) (id :: indices_towards_root) in
            let root, indices_on_path = loop child_ix parents.(child_ix) [] in
            change_parents_to_root parents indices_on_path root;
            root

        let union { parents; ranks } parent_index_p parent_index_q =
            if parent_index_p <> parent_index_q then
                if ranks.(parent_index_p) > ranks.(parent_index_q) then parents.(parent_index_q) <- parent_index_p
                else
                    parents.(parent_index_p) <- parent_index_q;
                    if ranks.(parent_index_p) = ranks.(parent_index_q)
                    then ranks.(parent_index_q) <- ranks.(parent_index_q) + 1
    end


let calc_distance { x = x1; y = y1 } { x = x2; y = y2 } = hypot (x1 -. x2) (y1 -. y2)


let read_two_floats() = Scanf.sscanf (read_line ()) "%f %f" (fun x y-> x, y)


let read_points nr_points =
    let rec loop acc k =
        if k = nr_points then acc
        else
            let x, y = read_two_floats() in
            loop ({ x; y } :: acc) (k + 1)
    in loop [] 0


let calc_pairwise_distances points =
    let distances =
        List.mapi
            (fun ix p ->
                List.mapi
                    (fun jy q -> { index_of_p = ix; index_of_q = ix + jy + 1; distance = calc_distance p q })
                    (List.drop (ix + 1) points))
        points in
    let comparison { distance = d1 } { distance = d2 } = Pervasives.compare d1 d2
    in List.sort comparison @@ List.concat distances


let calc_optimal_clustering nr_points points k =
    let distances = calc_pairwise_distances points in
    let clusters = UnionFind.initialize nr_points in
    let rec loop nr_clusters xs =
        let { index_of_p; index_of_q; distance } = List.hd xs in
        let cluster_of_p = UnionFind.find clusters index_of_p in
        let cluster_of_q = UnionFind.find clusters index_of_q in
        if cluster_of_p = cluster_of_q then loop nr_clusters (List.tl xs)
        else
            let () = UnionFind.union clusters cluster_of_p cluster_of_q in
            if nr_clusters = k then distance
            else loop (nr_clusters - 1) (List.tl xs)
    in loop nr_points distances


let () =
    let nr_points = read_int() in
    let points = read_points nr_points in
    let k = read_int() in
    let d = calc_optimal_clustering nr_points points k in
    print_float d;
    print_newline()
